packages/core/useWebLocks/index.ts

```diff
@@ -0,0 +1,113 @@
+  /**
+   * Either "exclusive" or "shared". The default value is "exclusive". Use "shared" to e.g. implement a one-writer-multiple-reader pattern.
+   *
+   * @default "exclusive"
+   */
+  mode?: LockMode
+
+  /**
+   * If `true`, the lock request will only be granted if it is not already held. If it cannot be granted, `onUnavailable` will be invoked.
+   *
+   * @default false
+   */
+  ifAvailable?: boolean
+
+  /**
+   * If `true`, then any held locks with the same name will be released, and the request will be granted, preempting any queued requests for it.
+   * This is an escape hatch to resolve deadlock situations. You should not need to use it.
+   *
+   * @default false
+   */
+  steal?: boolean
+
+  /**
+   * This function gets executed when `ifAvailable` is `true` and the lock could not get acquired immediately.
+   */
+  onUnavailable?: () => void
+}
+
+export interface UseWebLocksReturn<Keys extends string = string> {
+  /**
+   * Returns whether the Web Locks API is supported by the current browser.
+   */
+  isSupported: ComputedRef<boolean>
+
+  /**
+   * Request a web lock. Can be called simultaneously with different lock names to request different locks in parallel.
+   *
+   * @param name name of the lock to request. Cannot start with a hyphen (-). Needs to be unique in the whole origin.
+   * @param options optional options for the lock.
+   * @param callback function to call when the lock is held. It gets an abort signal as parameter that can be used to bail when the scope gets disposed while the lock is held.
+   * @return Promise that resolves when the lock is released or gets rejected when there was an error getting the lock.
+   */
+  request: ((name: Keys, callback: (signal: AbortSignal) => Promise<unknown> | unknown) => Promise<unknown>) & ((name: Keys, options: UseWebLocksRequestOptions, callback: (signal: AbortSignal) => Promise<unknown> | unknown) => Promise<unknown>)
+}
+
+/**
+ * Reactive Web Locks API. The lock automatically gets release on scope dispose.
+ *
+ * @see https://vueuse.org/useWebLocks/
+ * @param [options]
+ */
+export function useWebLocks<Keys extends string = string>(options: UseWebLocksOptions = {}): UseWebLocksReturn<Keys> {
+  const { navigator = defaultNavigator } = options
+  const isSupported = useSupported(() => navigator?.locks && typeof AbortController === 'function')
+  const releaseLocks = new Map<Keys, (value: unknown) => void>()
+  const abortControllers = new Set<AbortController>()
+  function request(name: Keys, callback: (signal: AbortSignal) => Promise<unknown> | unknown): Promise<unknown>
+  function request(name: Keys, options: UseWebLocksRequestOptions, callback: (signal: AbortSignal) => Promise<unknown> | unknown): Promise<unknown>
+  function request(name: Keys, optionsOrFn: UseWebLocksRequestOptions | ((signal: AbortSignal) => Promise<unknown> | unknown), callback?: (signal: AbortSignal) => Promise<unknown> | unknown): Promise<unknown> {
```

Comment:
i think we can have better types for `unknown` here

Reply:
How about this solution? library users can opt to use better types:
```ts
type ReturnTypeOne = void
type ReturnTypeTwo = { api: string, result: any }
const { request } = useWebLocks<{ lockName1: ReturnTypeOne, lockName2: ReturnTypeTwo }>()

request('lockName1', () => {})
request('lockName2', () => {
  // TypeScript error because api: number != api: string
  return { api: 1, result: 2}
})

```